ctools/defragment_sharded_collection.py

```diff
@@ -696,116 +687,127 @@ async def get_remain_chunk_imbalance(center, target_chunk):
                     continue
                 else:
                     break
+                
+            if total_shard_size[shard] * args.shard_imbalance_frac < max_shard_size and not center_size_kb == 0:
+                continue
 
             # chunks should be on other shards, but if this script was executed multiple times or 
             # due to parallelism the chunks might now be on the same shard            
 
             left_chunk = chunks_max_index.get(pickle.dumps(c['min']))
             right_chunk = chunks_min_index.get(pickle.dumps(c['max']))
             
-            # Exclude overweight target shards
-            if (left_chunk is not None and right_chunk is not None) and (left_chunk['shard'] != right_chunk['shard']):
-                if total_shard_size[left_chunk['shard']] > total_shard_size[right_chunk['shard']] * args.shard_imbalance_frac:
+            if left_chunk and right_chunk:
+                for dest_chunk in [left_chunk, right_chunk]:
+                    dest_chunk['votes'] = 0
+
+                    if dest_chunk['shard'] == shard:
+                        dest_chunk['votes'] += 1 << 5 
+                    dest_size = await get_chunk_size(dest_chunk)
+
+                    if center_size_kb <= dest_size:
+                        dest_chunk['votes'] += 1 << 4
+
+                    combined_size = center_size_kb + dest_size
+                    if combined_size > small_chunk_size_kb:
+                        if dest_size < small_chunk_size_kb:
+                            dest_chunk['votes'] += 1 << 3
+                        else:
+                            dest_chunk['votes'] += 1 << 1
+                    
+                if left_chunk['votes'] == right_chunk['votes']:
+                    if total_shard_size[left_chunk['shard']] < total_shard_size[right_chunk['shard']]:
+                        left_chunk['votes'] += 1 << 2
+                    elif total_shard_size[left_chunk['shard']] > total_shard_size[right_chunk['shard']]:
+                        right_chunk['votes'] += 1 << 2
+                
+                if right_chunk['votes'] > left_chunk['votes']:
                     left_chunk = None
-                elif total_shard_size[right_chunk['shard']] > total_shard_size[left_chunk['shard']] * args.shard_imbalance_frac:
-                    right_chunk = None
-                else:
-                    pass
+
 
             if left_chunk is not None:
```

Comment:
I believe that we could easily collapse in a single function the logic to move and merge the center with the chosen candidate (once you order the chunks by increasing minKey, the code for merging and updating the in-memory data structures becomes the same).

Reply:
I tried it and it is not exactly the same so I decided to leave it for another moment.

Reply:
How about something like this? (assumed to be a nested function)

```
    def move_chunk_and_merge_with_sibling(chunk, sibling, is_chunk_before_sibling):
        source_shard = chunk['shard']
        target_shard = sibling['shard']
        sibling_size = await get_chunk_size(sibling)
        new_size = sibling_size + center_size_kb
        first_merge_member = chunk if is_chunk_before_sibling else sibling
        second_merge_member = sibling if is_chunk_before_sibling else chunk
        merge_bounds = [first_merge_member['min'], second_merge_member['max']]
        if not args.dryrun:
            if source_shard != target_shard:
                await coll.move_chunk(chunk, target_shard)
            await coll.merge_chunks([first_merge_member, second_merge_member])
            if args.write_chunk_size:
                await coll.try_write_chunk_size(merge_bounds, target_shard, new_size)
        else:
            progress.write(f'Moving chunk [{chunk["min"]},{chunk["max"]}] from {shard} to {target_shard}, '
                            f'merging {merge_bounds}, new size: {fmt_kb(new_size)}')
        # update local map, 
        chunks_id_index.pop(second_merge_member['_id']) # only first chunk is kept
        chunks_min_index.pop(pickle.dumps(second_merge_member['min']))
        chunks_max_index.pop(pickle.dumps(first_merge_member['max']))
        chunks_max_index[pickle.dumps(second_merge_member['max'])] = first_merge_member
        first_merge_member['shard'] = target_shard
        first_merge_member['merged'] = True
        first_merge_member['max'] = c['max']
        first_merge_member['defrag_collection_est_size'] = new_size

        if source_shard != target_shard:
            total_shard_size[source_shard] -= center_size_kb
            total_shard_size[target_shard] += center_size_kb
            max_shard_size = max(max_shard_size, total_shard_size[target_shard]) 
            total_moved_data_kb += center_size_kb

        # update stats for merged chunk (source)
        progress.update(1)
        #update stats for merged chunk (destination)
        if sibling_size <= small_chunk_size_kb and new_size > small_chunk_size_kb:
            progress.update(1)

        await exec_throttle()
        begin_time = time.monotonic()
```